File: packages/cli/src/lib/address.ts
import { ProxerError } from "#app/lib/error.ts";

export type HostPort = {
  readonly host: string;
  readonly port: number;
};

export const parseHostPort = (
  input: string,
  defaultHost = "127.0.0.1",
): HostPort => {
  const separatorIndex = input.lastIndexOf(":");
  if (separatorIndex === -1) {
    throw new ProxerError("missing port");
  }

  const hostText = input.slice(0, separatorIndex).trim();
  const portText = input.slice(separatorIndex + 1).trim();
  if (portText.length === 0) {
    throw new ProxerError("missing port");
  }

  if (!/^\d+$/.test(portText)) {
    throw new ProxerError("invalid port");
  }

  const port = Number(portText);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new ProxerError("port must be between 1 and 65535");
  }

  return {
    host: hostText || defaultHost,
    port,
  };
};

export const formatHostPort = (address: HostPort): string => {
  return `${address.host}:${address.port}`;
};

File: packages/cli/lib/src/util/address.dart
import 'package:proxer/src/util/error.dart';

class HostPort {
  const HostPort(this.host, this.port);

  final String host;
  final int port;
}

HostPort parseHostPort(String input) {
  final value = input.trim();
  String host;
  String portText;

  if (value.startsWith('[')) {
    final end = value.indexOf(']');
    if (end < 0 || end + 1 >= value.length || value[end + 1] != ':') {
      throw ProxerError('address must be in host:port format');
    }
    host = value.substring(1, end);
    portText = value.substring(end + 2);
  } else {
    final separator = value.lastIndexOf(':');
    if (separator <= 0) {
      throw ProxerError('address must be in host:port format');
    }
    host = value.substring(0, separator);
    portText = value.substring(separator + 1);
  }

  final port = int.tryParse(portText);
  if (host.isEmpty || port == null || port < 0 || port > 65535) {
    throw ProxerError('address must be in host:port format');
  }
  return HostPort(host, port);
}

String formatHostPort(HostPort address) {
  final host = address.host.contains(':') ? '[${address.host}]' : address.host;
  return '$host:${address.port}';
}
